linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/VarGen.scala

```diff
@@ -232,6 +268,71 @@ private[emitter] final class VarGen(jsGen: JSGen, nameGen: NameGen,
     Ident(avoidClashWithGlobalRef("$" + field + "_" + subField), origName)
   }
 
+  private def maybeFromModule(className: ClassName, ident: Ident)(
+      implicit moduleContext: ModuleContext, globalKnowledge: GlobalKnowledge,
+      pos: Position): Tree = {
+    val targetModule = globalKnowledge.getModule(className)
+    if (targetModule == moduleContext.id) {
+      VarRef(ident)
+    } else {
+      DotSelect(VarRef(envModuleFieldIdent(targetModule.id)), ident)
+    }
+  }
+
+  private def maybeExport(ident: Ident, tree: Tree, mutable: Boolean = false)(
+      implicit moduleContext: ModuleContext, pos: Position): Tree = {
+    if (moduleContext.leaf) {
+      tree
+    } else {
+      config.moduleKind match {
+        case ModuleKind.NoModule =>
+          throw new AssertionError("non-leaf module in NoModule mode")
+
+        case ModuleKind.ESModule =>
+          // TODO: Mutable export!
+          val export = Export((ident -> ExportName(ident.name)) :: Nil)
+          Block(tree, export)
```

Comment:
@sjrd this is another thing we should discuss: Currently, we theoretically allow assigning to any static field from anywhere. Nothing I know of uses this. It is more difficult to do across ES6 modules because we need a dedicated setter function (and hence the trees look entirely different).

We can of course build it, but maybe only allowing to write to a static field from the context of the class it is defined in would be an acceptable squint-with-the-eyes IR change?

Otherwise, how do we test this :)

Reply:
Under optimizations, this is very difficult to ensure. You could have a method that modifies the static field in its class, but if that method is inlined in another class, the assignment will be moved to a different class, and there's not much you can do about it.

> Otherwise, how do we test this :)

With an `@inline def` :p

Reply:
Ah, right. Fair enough :)